### ISSUE TITLE
minibatch gradient

### ISSUE THREAD

**Issue by rajasekharponakala** [nitnelave/pycaffe_tutorial#2]
Hi, 

This is the awesome tutorial I've ever found!.

btw, Is there any way to find the gradients of loss function for a minibatch (computing adversarial examples), and train them and update to the network weights? using pycaffe API?

**Comment by nitnelave** [comment]
Hi,

Thanks for the comment, I appreciate it!

I'm not sure what you are trying to do. You can load a minibatch in the network, run the forward step, set the diff in the loss layer to what you want, then run the backward step to have the diff at the input layer. You can then apply that diff to your input to get closer to an adversarial example.

Keep in mind that I haven't played with Caffe for a while, so it might be a bit fuzzy, but that's the general idea.

**Comment by rajasekharponakala** [comment]
Ok. Thanks!

and with that diff, we can also update the weights, right?
like you mentioned in the 'network_manipulation script'

```
lr = 0.01
for l in net.layers:
    for b in l.blobs:
        b.data[...] -= lr * b.diff
```

**Comment by nitnelave** [comment]
Why do you want to update the weights? When generating adversarial examples, the network itself is usually not modified, because the examples are artificial and based only on the existing network.

**Comment by rajasekharponakala** [comment]
for example, I computed adversarial example by modifying the deploy network of googlenet.

```
force_backward: true
layer{
name: "data"
type: "Input"
top: "data"
input_param { shape: { dim: 1 dim: 1 dim: 224 dim: 224 } }
}
layer {
name: "label"
type: "Input"
top: "label"
input_param { shape: { dim: 1 dim: 1 dim: 1 dim: 1 } }
}
layer {
name: "conv1/7x7_s2"
type: "Convolution"
bottom: "data"
top: "conv1/7x7_s2"
param {
lr_mult: 1.0
decay_mult: 1.0
}
param {
lr_mult: 2.0
decay_mult: 0.0
}
.......
.......
.......
layer {
bottom: "loss3/classifier"
bottom: "label"
top: "loss"
name: "loss"
type: "SoftmaxWithLoss"
}

```

now i want to traing the same model to resist such attack. So to compute the adversarial example of a minibatch, we have to find the gradient of the loss fucntion for that minibatch, inorder to do this i have to add above similar layers to the `train.prototxt` right? is my view incorrect?

**Comment by nitnelave** [comment]
To train a network against adversarial attacks, I would generate many adversarial examples based on your training data, then train the network again using these examples (and the original ones).

**Comment by rajasekharponakala** [comment]
adversarial examples you mean the 'adding arbitrary pixel changes(noise) as a parameter of the network function' or the real-time pixel changes of the images. In the first case, both images look similar to the human eye, but the network thinks different because the pixels change according to their gradients with the network. In the second case, by using pre-processing techniques we add human seeable noise. Did you think the same?

**Comment by nitnelave** [comment]
An adversarial example is anything that a human would classify as something (ground truth) but that has been especially crafted so the network classifies it as something else. Whether the difference is human-perceptible or not is not really relevant (only changes the quality of the example).

**Comment by rajasekharponakala** [comment]
Thanks for the clarification. But then, how do you generate the adversarial example based on the training data?

**Comment by nitnelave** [comment]
See comment above: Run the network forward, update the objective, run the network backwards, get the gradient on the input layer, add it to the input, and repeat

**Comment by rajasekharponakala** [comment]
Thanks. I'll work on it. adding to input-  meaning the `input blob` or the values stored in the `npy` format ?
